spec/models/dynamic_field_group_spec.rb

```diff
@@ -0,0 +1,339 @@
+        let(:parent) { FactoryBot.create(:dynamic_field_category) }
+
+        context 'and there other top level groups' do
+          before do
+            FactoryBot.create(:dynamic_field_group, string_key: 'title', parent: parent, sort_order: 4)
+            FactoryBot.create(:dynamic_field_group, string_key: 'abstract', parent: parent, sort_order: 15)
+            parent.reload
+          end
+
+          it 'sets sort order to one more than the highest sort order of any top level group' do
+            expect(dynamic_field_group.sort_order).to be 16
+          end
+        end
+
+        context 'and there are no other top level groups' do
+          it 'sets sort order to 0' do
+            expect(dynamic_field_group.sort_order).to be 0
+          end
+        end
+      end
+
+      context 'and parent is a dynamic field group' do
+        let(:parent) { FactoryBot.create(:dynamic_field_group, string_key: 'all_names') }
+
+        context 'and other siblings are present' do
+          before do
+            FactoryBot.create(:dynamic_field_group, :child, parent: parent, sort_order: 9)
+            FactoryBot.create(:dynamic_field, string_key: 'type', dynamic_field_group: parent, sort_order: 16)
+            parent.reload
+          end
+
+          it 'sets sort order to one more than the higest sort order of any sibling' do
+            expect(dynamic_field_group.sort_order).to be 17
+          end
+        end
+
+        context 'and no other sublings are present' do
+          it 'sets sort order to 0' do
+            expect(dynamic_field_group.sort_order).to be 0
+          end
+        end
+      end
+    end
+
+    context 'when created_by is missing' do
+      let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, created_by: nil) }
+
+      it 'does not save' do
+        expect(dynamic_field_group.save).to be false
+      end
+
+      it 'returns correct error' do
+        dynamic_field_group.save
+        expect(dynamic_field_group.errors.full_messages).to include 'Created by must exist'
+      end
+    end
+
+    context 'when updated_by is missing' do
+      let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, updated_by: nil) }
+
+      it 'does not save' do
+        expect(dynamic_field_group.save).to be false
+      end
+
+      it 'returns correct error' do
+        dynamic_field_group.save
+        expect(dynamic_field_group.errors.full_messages).to include 'Updated by must exist'
+      end
+    end
+
+    context 'when xml_translation is invalid' do
+      let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, xml_translation: 'randomstring') }
+
+      it 'does not save' do
+        expect(dynamic_field_group.save).to be false
+      end
+
+      it 'returns correct error' do
+        dynamic_field_group.save
+        expect(dynamic_field_group.errors.full_messages).to include 'Xml translation does not validate as JSON'
+      end
```

Comment:
We should probably also add a validation rule someday that validates the the xml translation rules against a json schema for valid xml translation rule properties, but that can wait until later.  Maybe worth putting a TODO or "skip" test in here for that?

Reply:
That's a cool idea. I'll add a TODO